### ISSUE TITLE
README discussion

### ISSUE THREAD

**Issue by rarmatei** [CodeYourFuture/scot-graduation-project#1]
Here are the discussion points from the README - let's comment below

1. CSS: included is Bootstrap as CSS link. Is that ok? Using react-bootstrap (or react-material) has been a pain in the past.
2. CSS: Do we use something like CSS modules (it's easy, and might help us/them avoid cascading issues and work in paralle. Also another tickbox in their knowledge)
3. Divided the app into containers (for smart components with state, mostly top-level pages) vs components (dumb stateless components) folder. This is common practice, but maybe unnecessary for our scale?
4. Tests are moved to top-level __tests folder (we could introduce those, and have writing tests as stretch goal for some students)
5. use yarn or npm - which one to pick? It would be good to be consistent on server and client.
6. For the api calls, there is helpers/api.js module, do we put all api calls in one file or divide them, i.e. helpers/specific_api.js and helpers/specific_api_2.js. Feels like second approach might minimise conflicts?
7. What linting configuration to use?
8. Database creation - how do we do it? using pg library, where/how do we save the script to minimise conflicts? We agreed on droppping and recreating the DB everytime as opposed to doing migrations like last time to keep it simple.
9. Add linting, git hooks to project?
10. Add auth to skeleton, or at least, add step by step on how to implement it
11. Prepare walk-through of skeleton for project kickoff: most time spent on branching and Git flow process, but also cover high-level what happens on CI, introduce axios, project structure etc...

**Comment by rarmatei** [comment]
Here are my "rough" first thoughts:

1. Not sure. I guess Bootstrap is good? It doesn't require that much setup/pre-learning and it teaches them how to look through project docs. What problems were there with `react-material`? Can we use something like https://designrevision.com/docs/shards-react/getting-started so they have **core** components available and they also learn how to use 3rd party libs? It seems to be as easy as `import { Button } from "shards-react";`

2.  My vote would be for "no" initially - and then suggest it once they start having conflicts? (as it's something else they need to learn before starting to work on this)

3. Agree, I don't think it's necessary for these projects
4. Awesome!! Very good stretch goal!!
5. I say `npm` since it's a lot more popular
6. Yes, I think it's a good idea to separate to reduce conflicts
7. Doesn't CRA ship with linting? Can we use their defaults?
8. ehmmm I don't know
9. Related to 7 - but if we do decide to add our own - I can pick this up
10. I can create a video for this? And each team can follow it? Or should we bake it into this skeleton directly?

**Comment by llh1** [comment]
I just answer below when I have something to add compared to what Raresh said, otherwise, I agree with what is mentioned above.

1 - I'm also okay to include Bootstrap or any other css lib as it's really helpful to have a good looking app without spending too much time working on the css. I also didn't have good experience with React Bootstrap in the past, it's very heavy and their API was very unstable. I don't have experience with Shards-React.

5 -  +1 for npm, we haven't introduced yarn

6 - I guess it's something they should come up with, they should also feel the pain of conflicts and try to come up with a solution. Having mentors optimizing this kind of things from the very beginning may hide some learnings.

8 -  Let's just put a database.sql file at the root of the server folder. They can move it to another location if they want to.

10 -  A video or step by step workflow to implement it would be great rather than giving them a ready to use solution.

**Comment by bonboh** [comment]
1. Yes Bootstrap is fine. It's a good first CSS framework to use since it's the most popular.
2. No CSS modules to start with, like Rares said. I think it would be a good idea to promote namespaced CSS (like BEM) to students as this good CSS practise, and it should also lower conflicts since CSS classes are tied to specific components.
3. Agreed, unnecessary for this scale
4. Great idea of tests being a stretch goal so students can concentrate on practising what we taught in the course
5. npm since it's the most popular, and we don't want to introduce too many new things
6. I would do what is best practise in the real world (I don't know which is since I don't use express)
7. We can use eslint-config-standard? eslint-config-airbnb is great but is very restrictive and might create friction for students. Will we also use stylelint?
8. What Loic said
9. Yep use eslint (and stylelint?). Not sure what git hooks to use
10. What Rares said

**Comment by djgrant** [comment]
1. I like the look of https://material-ui.com/. https://react.semantic-ui.com/ also looks nice. I don't think it matters too much but I would like us to choose one which is stable and leverages React components well.
2. Ideally, we minimise custom CSS altogether and get students to lean on the UI framework. You get CSS modules for free with CRA though so it's not something we have to decide to include given it's already baked in. React Material UI uses an approach very similar to CSS modules, so it might be a good way to go if we use that framework. 
3. Let's definitely rename that `containers` folder to `components`. It would be great to if we can avoid bloated fat-controller-like components though. Maybe we could have a masterclass and show them how to refactor into one of these patterns after a few weeks.
4. I think it's much easier to co-locate tests with files. Integration tests only at the top level.
5. npm
6. Helpers is too ambiguous. This is a common structure:
- `routes` - express route handlers. Deal with getting/setting data and sending a response to the client. 
- `services` - semi-opionated interfaces for things like DB and APIs
- `utils` - small pure functions for things like transforming data 
7. I like @bonboh's suggestion.
8. Surely we can't just drop and recreate the DB once we've got prod data in it? 
9. Absolutely. Husky and lint-staged.

**Comment by kabaros** [comment]
I think there is pretty much agreement on those

2. No css modules to start with
3. Combine `containers` and `components` into just `components` #3 
4. Will move tests to the same level as the components #3 
5. use npm 
6. we will have a structure of `services` and `utils` eventually (not helpers) .. but let them get their themselves #3 
7. Use elint-config-standard #4
8. Use database.sql at the root level. This has been the easiest route in the last projects (when there is prod data, we can drop the dropping of database @djgrant ) #5
9. Add Husky and lint-staged #4

We need more discussions about these:

1. CSS - I lean towards using Bootstrap (not react-bootstrap), most of them still can't use it and it's the most common library. I found React-Material-UI  and 
10. Auth
